runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -7294,17 +7300,126 @@ done:;
+				U_32 rhsValue = _objectAccessBarrier.inlineMixedObjectReadU32(_currentThread, rhs, startOffset + result->offset);
+				I_32 result = helperFloatCompareFloat((jfloat *)&lhsValue, (jfloat *)&rhsValue);
+				if (result != 0) {
+					return false;
+				}
+				break;
+			}
+			case '[': { /* Array */
+				j9object_t lhsObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, lhs, startOffset + result->offset);
+				j9object_t rhsObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, rhs, startOffset + result->offset);
+				if (lhsObject != rhsObject) {
+					return false;
+				}
+				break;
+			}
+			case 'L': { /* Nullable class type or interface type */
+				j9object_t lhsObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, lhs, startOffset + result->offset);
+				j9object_t rhsObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, rhs, startOffset + result->offset);
+				if (!J9_IS_J9CLASS_VALUETYPE(J9OBJECT_CLAZZ(_currentThread, lhsObject))) {
+					if (NULL != lhsObject && NULL != rhsObject) {
+						if (lhsObject != rhsObject) {
+							return false;
+						}
+						break;
+					} else {
+						return false;
+					}
+				}
+				/* Fall through if we find this to be a value type at runtime */
+			}
+			case 'Q': { /* Null-free class type */
+				J9Class *fieldClass = findJ9ClassInFlattenedClassCache(clazz->flattenedClassCache, sigChar + 1, J9UTF8_LENGTH(signature) - 2);
+				bool recursiveResult = false;
+
+				if (J9_IS_J9CLASS_FLATTENED(fieldClass)) {
+					/* If J9ClassCanSupportFastSubstitutability is set, we can use memcmp, else we need to recursively check the fields manually. */
+					if (J9_ARE_ALL_BITS_SET(fieldClass->classFlags, J9ClassCanSupportFastSubstitutability)) {
+						void *lhsFieldStart = (void *)((UDATA)lhs + startOffset + result->offset);
+						void *rhsFieldStart = (void *)((UDATA)rhs + startOffset + result->offset);
+						if (0 != memcmp(lhsFieldStart, rhsFieldStart, J9_VALUETYPE_FLATTENED_SIZE(fieldClass))) { /* TODO: add GC write barrier check */
+							return false;
+						}
+						break;
+					} else {
+						recursiveResult = isSubstitutable(lhs, rhs, startOffset + result->offset, fieldClass);
+					}
+				} else {
+					/* When unflattened, we get our object from the specified offset, then increment past the header to the first field. */
+					j9object_t lhsFieldObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, lhs, startOffset + result->offset);
+					j9object_t rhsFieldObject = _objectAccessBarrier.inlineMixedObjectReadObject(_currentThread, rhs, startOffset + result->offset);
+					if (NULL != lhsFieldObject && NULL != rhsFieldObject) {
+						recursiveResult = isSubstitutable(lhsFieldObject, rhsFieldObject, J9VMTHREAD_OBJECT_HEADER_SIZE(_currentThread), fieldClass);
+					} else {
+						return false;
+					}
+				}
+
+				if (!recursiveResult) {
```

Comment:
this is not need, you can just return `recursiveResult`

Reply:
if there are multiple L types or Q types, do we not have to check them all?

Reply:
ah right, I see, you are still iterating